### PR TITLE
Feature/new packfile parser

### PR DESCRIPTION
**Pull request by jfontan**
Needs #896

It needs tests and index writer is still not complete:

* 64 bit offsets
* Index checksum

Some improvements can be done changing the cache to hold byte slices and index by offset. We can also skip reading headers in `resolveDeltas` phase.

### CODE REVIEW

File: plumbing/format/idxfile/writer.go [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+import (
+	"bytes"
+	"math"
+	"sort"
+
+	"gopkg.in/src-d/go-git.v4/plumbing"
+	"gopkg.in/src-d/go-git.v4/utils/binary"
+)
+
+type object struct {
```
Comment (erizocosmico) [inline plumbing/format/idxfile/writer.go]: we already have `Entry` in this package with this exact same struct, maybe you can use that?

File: plumbing/format/idxfile/writer.go [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+	hash   plumbing.Hash
+	offset int64
+	crc    uint32
+}
+
+type objects []object
```
Comment (erizocosmico) [inline plumbing/format/idxfile/writer.go]: are we using the `objects` type for something specific (i don't see any methods for it)? If not, I'd say just go with `[]object` (or `[]Entry`)

Comment (erizocosmico) [reply plumbing/format/idxfile/writer.go]: Nvm, saw the methods at the end of the file

**Comment by jfontan** [comment]
Now it supports 64 bit offsets but needs more testing. Index checksum is already generated by the encoder.

**Pull request merged**